### ISSUE TITLE
Possible error

### ISSUE THREAD

**Issue by kemin711** [pezmaster31/bamtools#145]
int tagval=-1;
When I use GetTag("NM", tagval);
The function is stuck at line: if ( !TagTypeHelper<T>::CanConvertFrom(type)
in BamAlignment.h

When I looked at the TagData it was like:  NMC^AMDZ98T2

const char type = *(pTagData - 1);

The type variable has a value of 'C' instead of 'i'.  I am using the latest version of bwa.  Looking into
the code of bwa, it is using the type i.  The MD string is properly typed as Z.  

Not sure this is caused by some modification from my part or not.  Any one can give a pointer as to how to fix this bug?  I will do the fixing.

**Comment by kemin711** [comment]
so far my temporary fix is like the following in the GetTag() function:

    if (type == 'C' && (tag == "NM" || tag == "AS" || tag == "XS")) { // patch a bug
       //type = Constants::BAM_TAG_TYPE_INT32;
      //cout << __FILE__ << ":" << __LINE__ << ": pTagData: "
      //   << pTagData << endl;
      //cerr << "NM int val at pTagData: " << (int)(*pTagData) << " at 2 bytes later: "
      //   << (int)(*(pTagData+2)) << endl;
       // this is a short term patch for the bug
      destination=(int)(*pTagData);
      return true;
    }
    else if ( !TagTypeHelper<T>::CanConvertFrom(type) ) {
        // TODO: set error string ?
       cerr << __FILE__ << ":" << __LINE__ << ":"
          << "Failed to convert to " << typeid(destination).name()
          << " from " << type << endl;
        return false;
    }

This is a patch,  not a real fix.

**Comment by angadps** [comment]
I used the GetTagType function (http://pezmaster31.github.io/bamtools/struct_bam_tools_1_1_bam_alignment.html#ac9cac21df9ef4a0a3414387b97cf14da) to identify the 'c' type (http://pezmaster31.github.io/bamtools/_bam_constants_8h.html) and ended up using int8_t to read the NM tag.

**Comment by kemin711** [comment]
Does the bam definition have C type?

**Comment by douglasgscofield** [comment]
Any progress on this?  I may be seeing the same error.  `samtools view` shows the read as:
````
ST-E00216:212:H5MHMCCXY:1:2208:25033:50832      99      NMWX01000001.1  99      60      7S144M  =       515     567     AATTCCAGACAGCGGATGCGCCTCCATTCGAGCACCAAATCGTTAAGGTCGTGGAACCTGCGGAGTGGGTGGGAAACTGCGTTAAAGTACCGGATGAAAAGTGGAAACGCTGGGTAAGTGTTTGAAAATGTGTTTCTCTTAAAACAATGGA        A<A<FJFFJJJJJFJJFJFJJJFA7AJJJJAAFAJJFFFJJFJ<J-FJJJJFJF7---<FFJJ<F-F777FJJJ-7F7JJFA7FAFJ-FJFFJFAA7J7AFJJJF-7FJJFFJF77-AJJF<J-7-<7FFFJA7F--AFF77-AF7F-7-<        MC:Z:151M      MD:Z:37G25T52T25A1      NM:i:4  MQ:i:60 AS:i:127        XS:i:19 RG:Z:QF-1499-T1-37_S2
````
and the Alignment is
````
Alignment = {
        Name = "ST-E00216:212:H5MHMCCXY:1:2208:25033:50832", Length = 151,
        QueryBases = "AATTCCAGACAGCGGATGCGCCTCCATTCGAGCACCAAATCGTTAAGGTCGTGGAACCTGCGGAGTGGGTGGGAAACTGCGTTAAAGTACCGGATGAAAAGTGGAAACGCTGGGTAAGTGTTTGAAAATGTGTTTCTCTTAAAACAATGGA",
        AlignedBases = "GACAGCGGATGCGCCTCCATTCGAGCACCAAATCGTTAAGGTCGTGGAACCTGCGGAGTGGGTGGGAAACTGCGTTAAAGTACCGGATGAAAAGTGGAAACGCTGGGTAAGTGTTTGAAAATGTGTTTCTCTTAAAACAATGGA",
        Qualities = "A<A<FJFFJJJJJFJJFJFJJJFA7AJJJJAAFAJJFFFJJFJ<J-FJJJJFJF7---<FFJJ<F-F777FJJJ-7F7JJFA7FAFJ-FJFFJFAA7J7AFJJJF-7FJJFFJF77-AJJF<J-7-<7FFFJA7F--AFF77-AF7F-7-<", TagData = "MCZ151M\000MDZ37G25T52T25A1\000NMC\004MQC<ASC\177XSC\023RGZQF-1499-T1-37_S2", RefID = 0,
````

Zooming in on the tags shows the read has:
```
MC:Z:151M      MD:Z:37G25T52T25A1      NM:i:4  MQ:i:60 AS:i:127        XS:i:19 RG:Z:QF-1499-T1-37_S2
```
which results in the Alignment containing
```
TagData = "MCZ151M\000MDZ37G25T52T25A1\000NMC\004MQC<ASC\177XSC\023RGZQF-1499-T1-37_S2"
```
I am seeing `C` in place of `i` and the `\023` separating it from `RGZ` (how I am seeing the bug) is equivalent to the decimal 19 (which follows `XS:i:`) in octal.

The same pattern appears for other tags and `:i:` in this read.  `NM:i:4` produces `NMC\004`, `MQ:i:60` produced `MQC<`, and `AS:i:127` produced `ASC\177`.

**Comment by douglasgscofield** [comment]
Yes, see Section 1.5 of <http://samtools.github.io/hts-specs/SAMv1.pdf>.  Integer types on tags (e.g., XS:i:20 in SAM format) may be encoded into the smallest integer that can hold the integer value.  In the case of XS:i:20, this was encoded as XSC\024.  The `C` indicates a uint8_t.  Use GetTagType() to figure out the tag type (for `C`, the value `Constants::BAM_TAG_TYPE_UINT8` will be returned) and then store it in an appropriate variable.
